Fix bugs in handling of negative slice + gather indices

jamesr66a opened the pull request:
Summary: This fixes multiple bugs in the handling of negative indices in both slicing and gather operations. These were uncovered by @[1466077526:Elias Ellison]'s diff D9493614, which made it so that we actually emit negative indices when we see them in PyTorch code.

Differential Revision: D9546183

houseroad reviewed (approved):
LG, could you also add some test cases in test_pytorch_onnx_caffe2.py? Also update the doc https://github.com/pytorch/pytorch/blob/master/caffe2/python/onnx/ONNXOpCoverage.md

houseroad commented:
expect files need to be updated. Also we need to make sure that it works on the cases which axis is higher than 2 (according to the doc, c2 gather has problem with higher axis)

houseroad reviewed (commented):
Can we check the axis, if it 0 or 1, we emit Gather, otherwise, we emit `Slice`and `Squeeze`.

houseroad commented on torch/onnx/symbolic.py:
```diff
@@ -412,10 +412,9 @@ def chunk(g, self, chunks, dim):
 def select(g, self, dim, index):
-    slice_node = g.op("Slice", self, axes_i=[dim], starts_i=[index], ends_i=[index + 1])
-    return g.op("Squeeze", slice_node, axes_i=[dim])
+    return g.op("Gather", self, index, axis_i=dim)
```
Can we check dim, if dim is 0 or 1, just emit `Gather`, otherwise use the old logic.

jamesr66a replied:
I think that should go in the caffe2 backend, since it's an implementation detail of caffe2 rather than something relating to the ONNX spec. But I can do it

houseroad commented:
@jamesr66a thanks for adding the fall back :-) agree, best solution is extending caffe2's kernels.

Commit:
Fix bugs in handling of negative slice + gather indices (#10973)

Summary:
This fixes multiple bugs in the handling of negative indices in both slicing and gather operations. These were uncovered by @[1466077526:Elias Ellison]'s diff D9493614, which made it so that we actually emit negative indices when we see them in PyTorch code.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10973

Reviewed By: jhcross

Differential Revision: D9546183

Pulled By: jhcross

fbshipit-source-id: 32b0362eee757e67c5a09144131b1d4eadb9f760
